In [35]:
import os
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [36]:
print(os.getcwd())

e:\spending_analyzer\Student_spending_analyzer\notebooks


In [37]:
##load the dataset

students=pd.read_csv("E:/spending_analyzer/Student_spending_analyzer/data/students.csv")
transactions=pd.read_csv("E:/spending_analyzer/Student_spending_analyzer/data/transactions.csv")

monthly_features=pd.read_csv("E:/spending_analyzer/Student_spending_analyzer/data/monthly_features.csv")


In [38]:
print(students.shape)
print(transactions.shape)
print(monthly_features.shape)

(300, 6)
(80000, 16)
(1200, 16)


In [39]:
transactions

transaction_id user_id account_id        date  \
0      2f3cd9ea-2611-4f15-8c1e-1d4da52c3c41  U00066     A00066  2025-08-01   
1      35cbff9c-1e43-4256-8ccc-6b0b44f11552  U00114     A00114  2025-08-02   
2      a5652d5f-8455-423b-8cfb-618e848372b3  U00101     A00101  2025-08-01   
3      de37568b-1a56-47d4-a27e-dda47b40e797  U00245     A00245  2025-08-01   
4      b61d3a54-2d82-4dbd-8948-9626153a3bfb  U00114     A00114  2025-08-01   
...                                     ...     ...        ...         ...   
79995  0325cc28-f820-4e54-a98c-ddbcd7e4dbcc  U00275     A00275  2025-11-13   
79996  7936284a-81dd-497d-bb20-ac9d37d95f35  U00188     A00188  2025-11-13   
79997  277af6e1-fee8-49c8-ae0e-4d9ed2f20f04  U00273     A00273  2025-11-13   
79998  e7a2d8fc-1140-4ce3-a840-7180a5acc10d  U00267     A00267  2025-11-13   
79999  f3a83141-760f-4c8f-8642-2b4a38c39ad4  U00205     A00205  2025-11-13   

                       timestamp    amount transaction_type       category  \
0      2025-08-01T08:01:07+05:30  21409.72           credit          other   
1      2025-08-02T06:51:18+05:30  -5755.80            debit           rent   
2      2025-08-01T06:09:59+05:30   -185.42            debit           food   
3      2025-08-01T06:17:43+05:30  -1682.95            debit       academic   
4      2025-08-01T06:00:16+05:30    -75.49            debit         travel   
...                          ...       ...              ...            ...   
79995  2025-11-13T10:57:40+05:30   -439.97            debit  entertainment   
79996  2025-11-13T10:59:12+05:30   -277.60            debit      groceries   
79997  2025-11-13T10:59:49+05:30    -81.37            debit         travel   
79998  2025-11-13T11:00:49+05:30   -181.39            debit           food   
79999  2025-11-13T11:02:20+05:30   -231.53            debit       academic   

              merchant  is_weekend  is_recurring recurring_id  balance_after  \
0      Family Transfer       False         False          NaN     -122971.14   
1           Hostel Fee        True          True  RENT_U00114       70230.57   
2       Campus Canteen       False         False          NaN       38701.17   
3              Lab Fee       False         False          NaN      155988.37   
4             Bus Pass       False         False          NaN       53766.73   
...                ...         ...           ...          ...            ...   
79995     Amazon Prime       False         False          NaN      -29359.26   
79996     Local Kirana       False         False          NaN      -54929.45   
79997       Local Auto       False         False          NaN       59900.31   
79998       Chai Point       False         False          NaN      109863.86   
79999  Stationery Shop       False         False          NaN        6793.25   

                    note data_quality_flag year_month  
0      Monthly allowance                ok    2025-08  
1           Monthly rent                ok    2025-08  
2                    NaN                ok    2025-08  
3                    NaN                ok    2025-08  
4                    NaN                ok    2025-08  
...                  ...               ...        ...  
79995                NaN                ok    2025-11  
79996       Manual entry                ok    2025-11  
79997       Manual entry                ok    2025-11  
79998                NaN                ok    2025-11  
79999                NaN                ok    2025-11  

[80000 rows x 16 columns]

In [40]:
transactions.transaction_id.nunique() #this means there are 3 duplicate transaction ids in the transactions dataset

79997

In [41]:
students

user_id accommodation  has_scholarship  has_internship     persona  \
0    U00001   shared_flat            False           False      normal   
1    U00002     hostel_pg            False           False      foodie   
2    U00003     hostel_pg            False            True      normal   
3    U00004     hostel_pg             True           False  bigspender   
4    U00005  with_parents             True           False      normal   
..      ...           ...              ...             ...         ...   
295  U00296     hostel_pg            False           False      foodie   
296  U00297         alone            False           False  bigspender   
297  U00298  with_parents            False           False      normal   
298  U00299  with_parents            False           False       saver   
299  U00300     hostel_pg            False           False       saver   

     expected_monthly_income  
0                   11311.28  
1                   22874.75  
2                   22902.65  
3                   19467.19  
4                   24871.38  
..                       ...  
295                 10966.76  
296                 17407.52  
297                 15832.28  
298                 10876.91  
299                 13907.75  

[300 rows x 6 columns]

In [42]:
monthly_features

user_id year_month  total_spend  total_income  spend_food  spend_rent  \
0     U00001    2025-08     28528.74      11311.28     4198.38     5432.92   
1     U00001    2025-09     20236.59      11311.28     2090.17     4190.97   
2     U00001    2025-10     25513.64      11311.28     3016.05     5049.40   
3     U00001    2025-11     16845.85      11311.28     2183.84     8471.23   
4     U00002    2025-08     33279.90      22874.75     8927.87     6284.27   
...      ...        ...          ...           ...         ...         ...   
1195  U00299    2025-11      2405.68      10876.91      209.13        0.00   
1196  U00300    2025-08     13401.21      13907.75      750.32     4590.61   
1197  U00300    2025-09     12227.12      13907.75      351.93     4237.84   
1198  U00300    2025-10     14736.92      13907.75      721.10     4186.89   
1199  U00300    2025-11      7656.83       8274.52      696.88     3690.72   

      spend_academic  spend_groceries  avg_daily_spend  median_transaction  \
0           10292.01          2486.92           950.96              220.02   
1            7765.88          1298.64           778.33              183.24   
2            5137.71          3106.34           981.29              208.14   
3            1917.54           561.42          1531.44              235.83   
4            9426.92          4004.54          1109.33              187.92   
...              ...              ...              ...                 ...   
1195          313.78           369.99           240.57              114.40   
1196         2283.26          1978.08           582.66              149.26   
1197         2285.03          1450.09           531.61              134.44   
1198         3326.75          2431.94           508.17              106.82   
1199         1201.64           952.15           850.76              227.06   

      max_transaction  num_transactions  weekend_spend_ratio  recurring_share  \
0             5432.92                90               0.2966           0.1904   
1             4190.97                69               0.3366           0.2071   
2             5049.40                82               0.2476           0.1979   
3             8471.23                36               0.7649           0.5029   
4             6284.27               108               0.1548           0.1888   
...               ...               ...                  ...              ...   
1195           409.56                17               0.3958           0.0000   
1196          4590.61                47               0.1669           0.3426   
1197          4237.84                47               0.4578           0.3466   
1198          4186.89                60               0.1633           0.2841   
1199          3690.72                21               0.2142           0.4820   

      savings_rate  label_anomalous  
0          -1.5221            False  
1          -0.7891            False  
2          -1.2556            False  
3          -0.4893            False  
4          -0.4549            False  
...            ...              ...  
1195        0.7788            False  
1196        0.0364            False  
1197        0.1208            False  
1198       -0.0596            False  
1199        0.0746            False  

[1200 rows x 16 columns]

In [43]:
print(students.info())
print("\n")
print(transactions.info())
print("\n")
print(monthly_features.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 6 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   user_id                  300 non-null    object 
 1   accommodation            300 non-null    object 
 2   has_scholarship          300 non-null    bool   
 3   has_internship           300 non-null    bool   
 4   persona                  300 non-null    object 
 5   expected_monthly_income  300 non-null    float64
dtypes: bool(2), float64(1), object(3)
memory usage: 10.1+ KB
None


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 80000 entries, 0 to 79999
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   transaction_id     80000 non-null  object 
 1   user_id            80000 non-null  object 
 2   account_id         80000 non-null  object 
 3   date               80000 non-null  object

## Observation no 1

We have 
- 300 users
- 80,000 transactions
- 1,200 monthly rows

1,200 / 300 = 4 months per user

dataset likely spans 4 months per student.

In [44]:
students["user_id"].nunique()

300

In [45]:
students["expected_monthly_income"].describe()

count      300.000000
mean     18859.290767
std       6349.319171
min       5017.820000
25%      14123.572500
50%      17971.695000
75%      22877.912500
max      50132.010000
Name: expected_monthly_income, dtype: float64

In [46]:
transactions["date"]=pd.to_datetime(transactions["date"])
transactions["timestamp"]= pd.to_datetime(transactions["timestamp"])
#Because behavior = time.
#And time must be handled correctly
transactions.dtypes

transaction_id                          object
user_id                                 object
account_id                              object
date                            datetime64[ns]
timestamp            datetime64[ns, UTC+05:30]
amount                                 float64
transaction_type                        object
category                                object
merchant                                object
is_weekend                                bool
is_recurring                              bool
recurring_id                            object
balance_after                          float64
note                                    object
data_quality_flag                       object
year_month                              object
dtype: object

In [47]:
#how many transactions per user 

transactions.groupby("user_id").size().describe()

count    300.000000
mean     266.666667
std       56.518345
min      168.000000
25%      230.000000
50%      262.500000
75%      291.000000
max      428.000000
dtype: float64

- Total users = 300 
- Mean transactions per user = 266.67, so on average each user has ~267 transactions.
- Check total transactions: 266.67 X 300=80,000 — this exactly matches our transactions table size, so the grouping result is consistent with the raw row count.
- Median = 262.5 is slightly below the mean, so the distribution is mildly right‑skewed (a few users with higher counts pull the mean up).
- Std = 56.52 gives a coefficient of variation approx 0.212 (std/mean), meaning per‑user counts vary by ~21% relative to the mean.
- Range = 168 → 428 — users at the low and high ends differ by ~2.6×, so there are meaningful extremes to investigate.


In [48]:
transactions["amount"].describe()

count    80000.000000
mean       -17.519584
std       2633.591485
min     -79848.290000
25%       -326.790000
50%       -186.040000
75%       -105.077500
max      50132.010000
Name: amount, dtype: float64

In [49]:
transactions["amount"].quantile([0.01, 0.05, 0.95, 0.99])

0.01    -3264.3568
0.05     -815.5120
0.95      -41.4700
0.99    15024.6200
Name: amount, dtype: float64

- 0.01 → -3264.36
1% of transactions have amount ≤ -3264.36. These are the extreme negative tail (largest outflows if negative means outflow).
- 0.05 → -815.51
5% of transactions have amount ≤ -815.51.
- 0.95 → -41.47
95% of transactions have amount ≤ -41.47. Equivalently, only 5% of transactions are greater than -41.47.
- 0.99 → 15024.62
99% of transactions have amount ≤ 15024.62. The top 1% of transactions are larger than 15024.62 (large positive values).


In [29]:
transactions["category"].value_counts()

category
food             21935
travel           12060
groceries        11864
academic          9631
entertainment     9195
other             9039
utilities         3055
ecommerce         2408
rent               813
Name: count, dtype: int64

In [30]:
transactions["is_recurring"].value_counts()


is_recurring
False    79187
True       813
Name: count, dtype: int64

In [31]:
transactions["data_quality_flag"].value_counts()
# If some rows are flagged as low quality: You may need to exclude them later.

data_quality_flag
ok    80000
Name: count, dtype: int64

In [50]:
transactions["transaction_type"].value_counts()

transaction_type
debit     78543
credit     1457
Name: count, dtype: int64

In [51]:
transactions.groupby("transaction_type")["amount"].describe()

count          mean          std       min       25%  \
transaction_type                                                           
credit             1457.0  16618.975896  7494.105778   3034.72  11134.04   
debit             78543.0   -326.132368   890.958957 -79848.29   -330.70   

                       50%        75%       max  
transaction_type                                 
credit            16145.22  21409.720  50132.01  
debit              -189.64   -109.145    -10.00


For anomaly detection about overspending,
we must ONLY analyze:

debit transactions

If we include credit:

- Income spikes will confuse the model

- Monthly sums will cancel out

- Spending behavior becomes distorted

In [52]:
#isolate expense behavior
#Keeps only debit rows
#Converts negative values to positive for easier interpretation

expenses = transactions[transactions["transaction_type"] == "debit"].copy()  #- creates a Boolean mask: True for rows where the column transaction_type equals "debit", False otherwise.
expenses["amount"] = expenses["amount"].abs()


In [54]:
expenses.head()

transaction_id user_id account_id       date  \
1  35cbff9c-1e43-4256-8ccc-6b0b44f11552  U00114     A00114 2025-08-02   
2  a5652d5f-8455-423b-8cfb-618e848372b3  U00101     A00101 2025-08-01   
3  de37568b-1a56-47d4-a27e-dda47b40e797  U00245     A00245 2025-08-01   
4  b61d3a54-2d82-4dbd-8948-9626153a3bfb  U00114     A00114 2025-08-01   
5  8c954cb0-09d1-4879-895e-041febb91acb  U00071     A00071 2025-08-01   

                  timestamp   amount transaction_type  category  \
1 2025-08-02 06:51:18+05:30  5755.80            debit      rent   
2 2025-08-01 06:09:59+05:30   185.42            debit      food   
3 2025-08-01 06:17:43+05:30  1682.95            debit  academic   
4 2025-08-01 06:00:16+05:30    75.49            debit    travel   
5 2025-08-01 06:11:24+05:30   146.70            debit      food   

         merchant  is_weekend  is_recurring recurring_id  balance_after  \
1      Hostel Fee        True          True  RENT_U00114       70230.57   
2  Campus Canteen       False         False          NaN       38701.17   
3         Lab Fee       False         False          NaN      155988.37   
4        Bus Pass       False         False          NaN       53766.73   
5             KFC       False         False          NaN       69425.66   

           note data_quality_flag year_month  
1  Monthly rent                ok    2025-08  
2           NaN                ok    2025-08  
3           NaN                ok    2025-08  
4           NaN                ok    2025-08  
5           NaN                ok    2025-08

In [55]:
expenses.shape

(78543, 16)

In [56]:
expenses["amount"].quantile([0.90, 0.95, 0.99])

0.90     565.7540
0.95     823.9480
0.99    3330.9642
Name: amount, dtype: float64

90th percentile ≈ 566

→ 90% of expenses are below ₹566

95th percentile ≈ 824

→ 95% are below ₹824

99th percentile ≈ 3,330

→ Top 1% jump sharply